File: scripts/sdk/register_case_runner.ts
import { KAMNClient, SDKError } from "../../packages/kamn-sdk/src/index.ts";

type ParsedArgs = {
  agentType: string;
  modelFamily: string;
  capabilities: string[];
};

function usage(): never {
  throw new Error(
    "usage: register_case_runner.ts --agent-type <value> --model-family <value> [--capability <value>]...",
  );
}

function parseArgs(argv: string[]): ParsedArgs {
  let agentType = "";
  let modelFamily = "";
  const capabilities: string[] = [];

  for (let index = 0; index < argv.length; ) {
    const arg = argv[index];
    switch (arg) {
      case "--agent-type":
        if (index + 1 >= argv.length) {
          usage();
        }
        agentType = argv[index + 1] ?? "";
        index += 2;
        break;
      case "--model-family":
        if (index + 1 >= argv.length) {
          usage();
        }
        modelFamily = argv[index + 1] ?? "";
        index += 2;
        break;
      case "--capability":
        if (index + 1 >= argv.length) {
          usage();
        }
        capabilities.push(argv[index + 1] ?? "");
        index += 2;
        break;
      default:
        usage();
    }
  }

  return { agentType, modelFamily, capabilities };
}

function sanitize(value: string): string {
  return value.replaceAll("\n", " ");
}

const parsed = parseArgs(process.argv.slice(2));
if (!parsed.agentType && !parsed.modelFamily && parsed.capabilities.length === 0) {
  usage();
}

const client = new KAMNClient();
try {
  const did = client.register(
    parsed.agentType,
    parsed.modelFamily,
    parsed.capabilities,
  );
  console.log("status=ok");
  console.log(`did=${did}`);
} catch (error: unknown) {
  if (error instanceof SDKError) {
    console.log("status=error");
    console.log(`error=${sanitize(error.message)}`);
  } else if (error instanceof Error) {
    console.log("status=error");
    console.log(`error=${sanitize(error.message)}`);
  } else {
    console.log("status=error");
    console.log("error=unknown error");
  }
}

File: scripts/sdk/register_case_runner.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import sys
from pathlib import Path
from typing import List

ROOT_DIR = Path(__file__).resolve().parents[2]
if str(ROOT_DIR) not in sys.path:
    sys.path.insert(0, str(ROOT_DIR))

from kamn_sdk import KAMNClient, SDKError


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--agent-type", required=True)
    parser.add_argument("--model-family", required=True)
    parser.add_argument("--capability", action="append", default=[])
    return parser.parse_args()


def sanitize(value: str) -> str:
    return value.replace("\n", " ")


def main() -> None:
    args = parse_args()
    capabilities: List[str] = list(args.capability)

    client = KAMNClient()
    try:
        did = client.register(args.agent_type, args.model_family, capabilities)
        print("status=ok")
        print(f"did={did}")
    except SDKError as error:
        print("status=error")
        print(f"error={sanitize(str(error))}")


if __name__ == "__main__":
    main()
